### ISSUE TITLE
Feature: Include Schema Name for all objects

### ISSUE THREAD

**Issue by hecon5** [joyfullservice/msaccess-vcs-addin#421]
In our DB, we have many "default" schema objects (`dbo`). The export of the objects seems to only include schema names if they're not `dbo`. 

I think it would be better to include the Schema names regardless. This will ensure consistency between the MSSMS script export tooling and this tool. 

Alternately, ability to specify whether the schema names are added or not would be nice.

**Comment by bclothier** [comment]
Just a FYI - `SourceTableName` property of the `DAO.TableDef` will contain the two part name and should be used to link an Access table with a SQL Server table. Note it is legal to have `abc.Foo` and `def.Foo` tables but in Access, they cannot be both `Foo` as specified via the `Name` property; one'll have to be renamed something else.

**Comment by joyfullservice** [comment]
@hecon5 - After further consideration, I think you are right. I like the simplicity and consistency of always including the schema name, and this matches what you see in MSSMS, and the default naming for exporting scripts. (Although a difference here is that I am using subfolders for object types, vs. appending the type as a suffix.) Thanks!

**Comment by hecon5** [comment]
I was thinking about this this weekend, and one thing we'd like to do is to be able to have the tool copy out the files and then use either a hook or manually copy the files out to the folder we've put the backend code into; this will allow us to also separately track the SQL when we just make a backend change, or change user details, etc. 

To do this, we'd also need to add the type name, and ensure they're consistent with what SSMS exports. 

I'm not 100% sure if this should be an option, or just do it the same way as SSMS for consistency, but to properly use this we would really like to have the exported object names match either way they've been exported.

**Comment by hecon5** [comment]
> I just don't want to spend to much time on the edge case of the edge case. 😄

Lol! For sure! I actually spent more time yesterday thinking about this than I should've instead of other things, but I think, generally, you're right. I looked into what's available with VSCode MSSQL addin as another alternate, and they also do folder based saving, so that seems to be the way to do it; caveat, they don't add the Schema name, but IMHO I prefer the schema name added as it makes portability slightly easier (and is the same method as SSMS). 

With all that said, I think using the same naming convention as SSMS / MSSQL addins for folder names would allow consistency, and not require much in the way of re-work when devs decide to migrate up to a full blown SQL management suite (/s here, they are all over the place in terms of features, and I like all of them and hate them all for different reasons / features that one has the others can't seemingly implement).

**Comment by joyfullservice** [comment]
Would you happen to have an example handy that lists the more standard names for those folders?

**Comment by hecon5** [comment]
# TL/DR

There's nothing "handy" out there...just a bunch of internet posts and musings about why there isn't, and the hand-written/typed notes of desperate devs trying to document their code...like us.

## Microsoft's Reference

[MSFT's sys.objects (Transact-SQL)](https://learn.microsoft.com/en-us/sql/relational-databases/system-catalog-views/sys-objects-transact-sql?view=sql-server-ver16) provides a decent list, but weirdly, the type names found there aren't what SSMS or MSSQL extension for VS Code...and **despite both being written by Microsoft, they differ in a few cases, too.**

## SO

We also found this SO Question [Interpreting type codes in sys.objects in SQL Server](https://stackoverflow.com/questions/2907049/interpreting-type-codes-in-sys-objects-in-sql-server). But it's from 2010, and doesn't really answer what the "standard" folder names should be. But it's something.

## SSMS / MSSQL

For version control, either of these work.

For rebuilding / creating a new DB from scratch based on these flat files, SSMS provides all the definitions and details you would need, MSSQL leaves out a few parts that cut out overall file count and some of the clutter, but to rebuild you'd need more pieces and a tad more work to get underway. I assume that using the MSSQL extension and the source files it exported, there's some fancy footwork under the hood to build them for you, but I've never tried (and don't intend to).

## Tables / Views / Stored Proceedures

Both SSMS and MSSQL Extension (MSSQL from here on out) appear to name these the same. My PR should show those.

## Triggers: 
Both SSMS and MSSQL Extension triggers appear to be attached to the table/view they affect when exporting.

## Users
I haven't found any consistency here in naming of folders or exported DDL files for permissions. This is the wild west. SSMS creates a file with `.User.SQL` appended to the end, and MSSQL tosses a file into a "Security" folder (along with other permissions), but doesn't name the file differently really.

## Permissions
SSMS attaches permissions to the object (usually), and MSSQL makes files named "Permission_N.SQL"; the MSSQL naming is ... not helpful. We use SSMS, and personally, I don't like how MSSQL separates permissions nor how it names them because we've discovered that if you remove / add permissions they can all get renumbered...super helpful with change control. 

## Schema
SSMS prepends the schema name to the file name, and a separate Schema named file with `.Schema.SQL` suffix defining the schema itself. I find it helpful, and because we have a few views that are named the same for more than one schema (schema dependent permissions allows more / fewer rows to be visible), it makes tracking those easier (for our workflow).

MSSQL seems to just separate the schema files out, and not define the schema. 

# Conclusion
All this to say, there's nothing "handy", but what we made up internally while testing.

We don't have any functions, so I can't test those.

# MSSQL Exported Folder Names
This is MSSQL Export from VSCode. We just started trying it out within the last week or so (timing of this is impressive), so we don't have much experience in it, just the confused attempts of a neophyte clicking all the buttons at once.
![image](https://github.com/joyfullservice/msaccess-vcs-addin/assets/54177882/f431887f-14d7-40e0-a94b-5ddb714cf8e4)

# SSMS Exported Names
Since we don't have a lot of these, I'm not 100% certain of this, but for the objects we do have in these list, SSMS seems to adhere fairly closely to this for naming. The export appears to append the shown folder name, removing dashes and spacing, and the trailing "s" when exporting. 

Since we don't have any functions in the DB I have handy, I am only presuming those would follow similar naming as the rest of these. Hopefully someone else knows.

### SSMS Example names 
|Object Type::|Extension|Example Name|
|:-|-|-|
|Table|`Table.SQL`|`dbo.tblSomething.Table.SQL`|
|View|`View.SQL`|`dbo.qrySomethingNeat.View.SQL`|
|Stored Proceedure|`StoredProcedure.SQL`|`dbo.SomeStoredProc.StoredProcedure.SQL`|
|User def. Table Type|`UserDefinedTableType.SQL`|`dbo.SomeTableTypeName.UserDefinedTableType.sql`|

and so on

Here's the folder names we see in our instance of SSMS:
![image](https://github.com/joyfullservice/msaccess-vcs-addin/assets/54177882/3639e7b5-114c-497b-90c3-aa42531336d6)

**Comment by bclothier** [comment]
I don't see any mention of SQL Server Data Tools & database project. I would suggest looking at it since I believe this is what Microsoft would recommend for source control for the schema. Note that it's now a component of Visual Studio though it used to be a standalone tool but is available for free.

**Comment by hecon5** [comment]
> I don't see any mention of SQL Server Data Tools & database project. I would suggest looking at it since I believe this is what Microsoft would recommend for source control for the schema.
I actually had never heard of this product before in looking around on managing a SQL server; we also don't have VS, so that may be a function of that, too.

Do you know where they define the schema/naming conventions? I looked on the [SQL Server Data Tools](https://learn.microsoft.com/en-us/sql/ssdt/sql-server-data-tools?view=sql-server-ver16) online docs, but don't see any mention of naming conventions anywhere, mostly just that you can use it to compare a DB and other tools, but nothing about version control; there's also the chance I just scrolled right past it, too.

**Comment by hecon5** [comment]
I did find this [SO Thread asking basically the same question about naming:](https://dba.stackexchange.com/a/130588); it looks like MSSQL may use the same/similar folder naming as SSDT...may even be the same thing under the hood?

**Comment by bclothier** [comment]
I also don't think they necessarily make all folders; only if there's an object.
![image](https://github.com/joyfullservice/msaccess-vcs-addin/assets/2367644/9f27f1eb-bc14-4e90-bb8c-2e10a626b78e)

I also think there's an option to change the preferences but can't find ATM. 

Yes, that is how you version control it -- you would build or export a database project into a database, in very similar manner you would do with this add-in for Access code because the database project has all SQL files laid out allowing you to recreate the database. Of course, there are several settings to customize what should be scripted out/compared/imported/exported.

**Comment by joyfullservice** [comment]
I think we have solved the original issue of including the schema name for all objects. ✔️ Short of a widely-used standard for the folder names, I am content to just use the existing names that are more consistent with the standard folders created by this add-in. For those that need more flexibility or more advanced use cases, some of the other tools may be a better fit.

My main goal for external database schemas was to provide something basic that works right out of the box with minimal setup and configuration. I have a hunch that many Access developers don't do _any_ version control on their back-end SQL objects, and this _something_ should be a lot better than _nothing_.  😄

**Comment by hecon5** [comment]
I think you're right. Perfect is the enemy of done, and I think after thinking on it, you're right. If it becomes an issue, I can always make a PR to include those bits. Closing.
